src/lib/zod/index.ts

```diff
@@ -5,6 +5,26 @@ import {
   zStringISODateUTC,
 } from '../../../src/forms/formHelpers';
 
+const PERSONAL_SIGN_REGEX = /0x[0-9a-f]{130}/;
+export const loginInput = z.object({
+  address: zEthAddressOnly,
+  data: z.string().refine(
+    msg => {
+      const templateOk = msg.startsWith('Login to Coordinape');
+      const timestamp = msg.split(' ').pop();
+      const timestampInPast =
+        timestamp && Number.parseInt(timestamp) * 1000 < Date.now();
+
+      const validLength = msg.length === 30;
+
+      return templateOk && validLength && timestampInPast;
+    },
+    { message: 'Invalid message payload' }
+  ),
+  hash: z.string().length(0),
```

Comment:
the hash would have a value if its a smart contract wallet

Reply:
Ah thanks, didn't know we supported this.

Reply:
> Ah thanks, good point. Maybe we'll implement a handler for this when there's a use-case for smart contracts delegating others to obtain api keys?

Its currently in use though, initially we didn't have this, until the users that were using special wallets like [Argent](https://www.argent.xyz/) were complaining they couldn't log into our app using their wallet via walletconnect. So we added this part.

Reply:
I'm adding this now, since it's so straightforward

Reply:
> I'm adding this now, since it's so straightforward

@topocount 
theres a helper function i created that you might be able to reuse
https://github.com/coordinape/coordinape/blob/main/src/utils/provider.ts#L40

Reply:
@zashton sorry I missed this. I reimplemented this in the api-lib/sigatures.ts file a bit differently.